### PR TITLE
Misc minor enhancements

### PR DESCRIPTION
**Pull request by alyst**
Several changes that looked too small and innocent for a separate PR:
  * more tests for corner-case joins/groupings/assignments
  * implement `AbstractVector` interface for `Cols` (mostly required to allow `map!(..., columns(df))` for `SubDataFrame`)
  * some optimizations for `GroupedDataFrame` (avoiding array reallocation, improving type inference)

### CODE REVIEW

File: src/abstractdataframe/reshape.jl [hunk 1]
```diff
@@ -78,26 +78,24 @@ function stack(df::AbstractDataFrame, measure_vars::Vector{Int}, id_vars::Vector
...
-function stack(df::AbstractDataFrame, measure_vars::Int, id_vars::Vector{Int})
-    stackdf(df, [measure_vars], id_vars)
+function stack(df::AbstractDataFrame, measure_var::Int, id_vars::Vector{Int})
+    stackdf(df, [measure_var], id_vars)
 end
 stack(df::AbstractDataFrame, measure_vars, id_vars) =
     stack(df, index(df)[measure_vars], index(df)[id_vars])
-function stack(df::AbstractDataFrame, measure_vars)
+# no vars specified, by default select only numeric columns
+numeric_vars(df::AbstractDataFrame) = [T <: AbstractFloat || (T <: Nullable && eltype(T) <: AbstractFloat)
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl]: Numeric corresponds to `Number`, doesn't it? I don't know why the original code did that.

File: src/groupeddataframe/grouping.jl [hunk 2]
```diff
@@ -203,15 +203,14 @@ Not meant to be constructed directly, see `groupby` abnd
...
+    vals::Vector{T}
 
-    function GroupApplied(gd, vals)
-        if length(gd) != length(vals)
-            error("GroupApplied requires keys and vals be of equal length.")
-        end
-        new(gd, vals)
+    @compat function (::Type{GroupApplied})(gd::GroupedDataFrame, vals::Vector)
+        length(gd) == length(vals) ||
+            throw(DimensionMismatch("GroupApplied requires keys and vals be of equal length."))
```
Comment (nalimilan) [inline src/groupeddataframe/grouping.jl]: Could you add `(got $(length(gd)) and $(length(vals))` to the message?

File: src/groupeddataframe/grouping.jl [hunk 3]
```diff
@@ -203,15 +203,14 @@ Not meant to be constructed directly, see `groupby` abnd
...
+        new{eltype(vals)}(gd, vals)
```
Comment (nalimilan) [inline src/groupeddataframe/grouping.jl]: Probably more idiomatic (and maybe more efficient too) to make the eltype a parameter of the method.

File: src/abstractdataframe/abstractdataframe.jl [hunk 4]
```diff
@@ -197,14 +201,7 @@ eltypes(df)
...
 """
-function eltypes(df::AbstractDataFrame)
-    ncols = size(df, 2)
-    res = Array(Type, ncols)
-    for j in 1:ncols
-        res[j] = eltype(df[j])
-    end
-    return res
-end
+eltypes(df::AbstractDataFrame) = map!(eltype, Vector{Type}(size(df,2)), columns(df))
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl]: Wouldn't it be more readable as a comprehension? And isn't type inference able to choose the best element type?

File: src/groupeddataframe/grouping.jl (hunk 3 above)
Comment (alyst) [inline src/groupeddataframe/grouping.jl]: I thought that type parameters create some overhead for dispatch. Here there are no inter-argument type constraints, so it's not necessary. In terms of type inference and the resulting code both versions should be equivalent. But I can change it, if using type parameters is the recommended way.

File: src/abstractdataframe/abstractdataframe.jl (hunk 4 above)
Comment (alyst) [inline src/abstractdataframe/abstractdataframe.jl]: I have tried several variants, including `map()` and comprehensions. Unfortunately, `eltype()` returns `Any`, so it results in "reds" for `@code_warntype eltypes(..)`. The proposed version gives the shortest "no-reds" code.

File: src/abstractdataframe/reshape.jl (hunk 1 above)
Comment (alyst) [inline src/abstractdataframe/reshape.jl]: I can rename it, but I don't have a better name idea. I agree that the default behaviour of stack() is somewhat inconsistent with e.g. joins requiring all joining columns to be specified, and I would rather remove this default. Just made the change to squeeze some duplicated code.

File: src/abstractdataframe/abstractdataframe.jl (hunk 4 above)
Comment (ararslan) [inline src/abstractdataframe/abstractdataframe.jl]: So `Type[eltype(col) for col in columns(df)]` has element type `Any`, even with the type annotation in the front?

Comment (alyst) [inline src/abstractdataframe/abstractdataframe.jl]: `eltype(col)` has result type `Any`, and IIRC `eltype(col)::Type` doesn't help getting the type of array inferred automatically. Probably it would be as fast as `map!()`, I just wanted to have a version without any `Any` :)

File: src/groupeddataframe/grouping.jl (hunk 3 above)
Comment (nalimilan) [inline src/groupeddataframe/grouping.jl]: No idea...

File: src/abstractdataframe/reshape.jl (hunk 1 above)
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl]: Agreed, let's deprecate this later.

File: src/groupeddataframe/grouping.jl (hunk 2 above)
Comment (alyst) [inline src/groupeddataframe/grouping.jl]: done

**Comment by alyst** [comment]
@nalimilan Thanks for the review. I've just added fix for the `groupby()` doc. With this PR it would be easier for me to rebase #850

**Comment by nalimilan** [comment]
Should we merge this now?

**Comment by alyst** [comment]
@nalimilan Yes, if nobody objects.
